RavenDB-11352 Pull Replication

karmeli87 opened the pull request:
In pull replication there are two roles: central and edge.

In the central cluster we define on the database record a pull replication task with all the relevant settings like mentor, delay and certificates.
Once defined the central will wait for a connection to be initiated by an edge.

On the edge node we define an external replication with pull replication settings.
This means that a node in an edge cluster will request the tcp info for some node in the central and initiate the TCP connection so the central will send its documents via this connection.

Security:
The definition on the central will contain thumbprints for allowed certificates.
The edge need to put a certificate (with the private key) that corresponds to the thumbprint of the central definition. This certificate is store with different prefix so we wouldn't trust it by mistake.
Initiating a pull replication will use the edge certificate which the central should allow using the stored thumbprint in the corresponded definition.

karmeli87 committed:
RavenDB-11352 Pull Replication

In pull replication there are two roles: central and edge.

In the central cluster we define and keep on the database record a pull replication task with all the relevant settings like mentor, delay and certificates.
Once defined the central will wait for a connection to be initiated by the edge.

On the edge node we define an external replication with pull replication settings.
This means that a node in an edge cluster will request the tcp info for some node in the central and initiate the TCP connection so the central will send the document via this connection.

Security:
The definition on the central will contain thumbprints for allowed certificates.
The edge need to put a certificate (with the private key) that corresponds to the thumbprint of the central definition. This certificate is store with different prefix then the regular client certificates so we don't trust it by mistake.
Initiating a pull replication will use the edge certificate which the central should allow using the stored thumbprint in the corresponded definition.

ayende commented on src/Raven.Client/Documents/Operations/Replication/FeatureTaskDefinition.cs:
```diff
@@ -0,0 +1,53 @@
+{
+    public abstract class FeatureTaskDefinition : IDynamicJsonValueConvertible
+    {
+        public List<string> Certificates; // thumbprint
```
We want to hold the full cert here, so we can do things like show the expiration

karmeli87 replied:
Done.

ayende commented on src/Raven.Client/Documents/Operations/Replication/ExternalReplication.cs:
```diff
@@ -105,4 +108,31 @@ public string GetTaskName()
+    public class PullReplicationAsEdgeSettings
+    {
+        public bool PullReplicationEnabled = true;
+        public string CertificateThumbprint;
```
This is the definition on the _edge_, right? How do you decide which certificate to use here?

karmeli87 replied:
The user should define it.

ayende replied:
Sort of. If the user specified a certificate to use here, you'll use that, otherwise you'll use the usual server certificate as normal.

ayende commented on src/Raven.Client/ServerWide/Operations/Certificates/PutFeatureCertificateOperation.cs:
```diff
@@ -0,0 +1,73 @@
+                            writer.WriteString(_name.ToString());
+                            writer.WriteComma();
+                            writer.WritePropertyName(nameof(CertificateDefinition.Certificate));
+                            writer.WriteString(Convert.ToBase64String(_certificate.Export(X509ContentType.Pkcs12)));
```
Unless I miss something, this stores the _private_ key inside the server, that is _not_ something that we usually want to do.
Why is this done here?

ayende replied:
Especially when you are talking about the global cluster level?

karmeli87 replied:
This is a client code to upload the certificate along with it private key to the edge.
Storing the certificate is done at the endpoint.

ayende commented on src/Raven.Client/ServerWide/Operations/Certificates/PutFeatureCertificateOperation.cs:
```diff
@@ -0,0 +1,73 @@
+                            writer.WriteString(_name.ToString());
+                            writer.WriteComma();
+                            writer.WritePropertyName(nameof(CertificateDefinition.Certificate));
+                            writer.WriteString(Convert.ToBase64String(_certificate.Export(X509ContentType.Pkcs12)));
```
```suggestion
                            writer.WriteString(Convert.ToBase64String(_certificate.Export(X509ContentType.Cert)));
```

ayende commented on src/Raven.Client/ServerWide/Tcp/TcpConnectionHeader.cs:
```diff
@@ -18,6 +19,28 @@ public enum OperationTypes
+
+            public AuthorizeMethod AuthorizeAs;
+
+            public string RemoteConnectionInfo;
```
What is this string?

karmeli87 replied:
This is remote pull replication so, renamed to `PullReplicationDefinitionName`.

ayende replied:
Which we'll need to change when we have pull ETL?

karmeli87 replied:
I'm bad with names... how does `FeatureDefinitionName` sounds.

ayende replied:
`AuthorizationFor` is probably better.

ayende commented on src/Raven.Server/Commercial/LicenseManager.cs:
```diff
@@ -1320,6 +1320,19 @@ public void AssertCanAddExternalReplication()
+        public void AssertCanAddPullReplication()
+        {
+            //TODO: add this feature to the license
+//            if (IsValid(out var licenseLimit) == false)
```
@grisha-kotler What is required here?

ayende replied:
@grisha-kotler Please check this. http://issues.hibernatingrhinos.com/issue/RavenDB-12415

ayende commented on src/Raven.Server/Documents/Replication/OutgoingReplicationHandler.cs:
```diff
@@ -75,9 +77,14 @@ public class OutgoingReplicationHandler : IDisposable
-        public OutgoingReplicationHandler(ReplicationLoader parent, DocumentDatabase database, ReplicationNode node, bool external, TcpConnectionInfo connectionInfo)
+        // In case this is an outgoing pull replication from the central
+        // we need to associate this instance to the replication definition. 
+        public string PullReplicationDefinitionName;
```
What happen if we delete / disable the pull replication / change its settings when the connection is already alive?

ayende replied:
Stilling missing handling for this

ayende commented on src/Raven.Server/Documents/Replication/OutgoingReplicationHandler.cs:
```diff
@@ -118,237 +134,156 @@ public string GetNode()
+        private void PullReplication()
         {
-            try
+            NativeMemory.EnsureRegistered();
```
Pretty sure that this is already handled on `PoolOfThreads.GlobalRavenThreadPool.LongRunning`

ayende replied:
remove this line

ayende commented on src/Raven.Server/Documents/Replication/OutgoingReplicationHandler.cs:
```diff
@@ -118,237 +134,156 @@ public string GetNode()
-                        catch (OperationCanceledException e)
+            if (connected == false)
+            {
+                throw new InvalidOperationException("Failed to connect to any of the supplied TCP connections.");
```
Bad idea here, We need to keep track of all the failures that we had an use them in an AggregateException as the inner exception here.

ayende commented on src/Raven.Server/Documents/Replication/OutgoingReplicationHandler.cs:
```diff
@@ -118,237 +134,156 @@ public string GetNode()
+                throw new InvalidOperationException("Failed to connect to any of the supplied TCP connections.");
+            }
+
+            using (Interlocked.Exchange(ref _tcpClient, task.Result))
```
Why interlocked here?

ayende commented on src/Raven.Server/Documents/Replication/OutgoingReplicationHandler.cs:
```diff
@@ -118,237 +134,156 @@ public string GetNode()
+            var request = new DynamicJsonValue
+            {
+                ["Type"] = nameof(ReplicationInitialRequest),
+                [nameof(ReplicationInitialRequest.PullReplication)] = isPullReplication
```
Why do we need this is we have all the fields that are there just for pull replication?

ayende commented on src/Raven.Server/Documents/Replication/ReplicationLoader.cs:
```diff
@@ -160,7 +164,108 @@ public ReplicationLoader(DocumentDatabase database, ServerStore server)
+                            pullReplicationDefinition = _server.Cluster.ReadPullReplicationDefinition(Database.Name, initialRequest.PullReplicationDefinition, ctx);
+                        }
+
+                        var taskId = (int)Hashing.XXHash64.Calculate(initialRequest.DatabaseGroupId, Encodings.Utf8);
```
Don't use `(int)` to get an integer here, use `Hashing.Mix` to make sure we include all the bits in the random effect

ayende replied:
Also, why do you need the `taskId` here anyway?

karmeli87 replied:
This is converted to `int` anyway in the `WhosTaskIsIt`.
We need this since this is a legit external replication and we keep its state across the central cluster. 
https://github.com/ravendb/ravendb/blob/ab8b050739d768eab847dfc3e60937bee07cff46/src/Raven.Server/Documents/Replication/OutgoingReplicationHandler.cs#L260

ayende replied:
Why do you need to keep its state on the central cluster?  You may have many of them, so you don't really have _state_ here.
The problem isn't the int, the problem is the cast throwing away 50% of the bits without mixing them.

ayende commented on src/Raven.Server/Documents/Replication/ReplicationLoader.cs:
```diff
@@ -160,7 +164,108 @@ public ReplicationLoader(DocumentDatabase database, ServerStore server)
+        public void RunPullReplicationAsEdge(TcpConnectionOptions tcpConnectionOptions, JsonOperationContext.ManagedPinnedBuffer buffer, ReplicationNode destination)
+        {
+            var newIncoming = CreateIncomingReplicationHandler(tcpConnectionOptions, buffer);
+            //TODO: Incoming need to add, known new value
```
I don't understand this comment?

ayende replied:
You mean `_incoming.GetOrAdd` ? It is _not_ a known new instance, you might already have this instance live already

ayende commented on src/Raven.Server/Documents/Replication/ReplicationLoader.cs:
```diff
@@ -715,11 +803,26 @@ private TcpConnectionInfo GetConnectionInfo(ReplicationNode node, bool external)
+                        }
+                    }
+                   
+                    using (var requestExecutor = RequestExecutor.CreateForFixedTopology(exNode.ConnectionString.TopologyDiscoveryUrls, exNode.ConnectionString.Database, certificate, DocumentConventions.Default))
```
Can save some code by unifying the `requestExecutor ` and `ctx` usages across the `if`s

ayende commented on src/Raven.Server/Documents/Replication/ReplicationLoader.cs:
```diff
@@ -747,6 +853,38 @@ private TcpConnectionInfo GetConnectionInfo(ReplicationNode node, bool external)
+            if (exNode == null)
+                return _server.Server.Certificate.Certificate;
+
+            if (exNode.PullReplicationAsEdgeSettings?.CertificateThumbprint != null)
```
No, we don't want this.
Instead, we want to have a `Certificate` property on the `ExternalReplication` directly that would hold the BASE64 certificate (including private key)

ayende replied:
Actually, better would be to hold the actual certificate elsewhere (because the certificate private key is... private, but we need:
* To upload the actual pull certificate when creating the pull replication
* To prevent it from being downloaded

ayende commented on src/Raven.Server/Documents/Replication/ReplicationLoader.cs:
```diff
@@ -747,6 +853,38 @@ private TcpConnectionInfo GetConnectionInfo(ReplicationNode node, bool external)
+                using (_server.ContextPool.AllocateOperationContext(out TransactionOperationContext context))
+                using (context.OpenReadTransaction())
+                {
+                    var key = Constants.Certificates.FeaturePrefix + exNode.PullReplicationAsEdgeSettings.CertificateThumbprint;
```
You have no way here to limit this for a single db. In particular, just knowing the certificate thumbprint allows you to use it from any db on the server, which is probably not ideal

karmeli87 replied:
Okay, so when putting a certificate on the edge we have to specify permissions.

ayende replied:
No, you don't.
When you update a pull replication task you can provide it with a certificate, that will be stored internally (encrypted) and used.
No reuse between pull replications

ayende commented on src/Raven.Server/Documents/Replication/ReplicationLoader.cs:
```diff
@@ -747,6 +853,38 @@ private TcpConnectionInfo GetConnectionInfo(ReplicationNode node, bool external)
+                    authorizationInfo = new TcpConnectionHeaderMessage.AuthorizationInfo
+                    {
+                        AuthorizeAs = TcpConnectionHeaderMessage.AuthorizationInfo.AuthorizeMethod.PullReplication,
+                        RemoteConnectionInfo = exNode.PullReplicationAsEdgeSettings.RemoteName
```
What is this remote info?

karmeli87 replied:
renamed to `PullReplicationDefinition`. (note: this is the definition name on the central)

ayende replied:
This is something that we'll want to use for pull ETL as well

ayende commented on src/Raven.Server/Web/Authentication/AdminCertificatesHandler.cs:
```diff
@@ -261,12 +269,18 @@ public async Task Put()
 
             if (string.IsNullOrEmpty(password))
-                collection.Import(certBytes, (string)null, X509KeyStorageFlags.MachineKeySet);
+                collection.Import(certBytes, (string)null, X509KeyStorageFlags.MachineKeySet | X509KeyStorageFlags.Exportable);
```
why this change?

karmeli87 replied:
Need to extract the private key later. (see comment below)

ayende commented on src/Raven.Server/Web/Authentication/AdminCertificatesHandler.cs:
```diff
@@ -296,12 +310,16 @@ public static async Task PutCertificateCollectionInCluster(CertificateDefinition
                 {
-                    // avoid storing the private key
+                    // here we need to keep the private key, since in pull replication we are the initiator of the communication.
+                    currentCertDef.Certificate = Convert.ToBase64String(x509Certificate.Export(X509ContentType.Pfx));
```
This shouldn't be using the same mechanism, as certificates that we store that aren't private information.
This should be _encrypted_, talk with @iftahbe on how we store sensitive information like that.
I would also like it to be completely separated so you can't re-use existing value just by knowing its thumbprint.

karmeli87 replied:
Okay, so the encryption will be locally on each node, but I need to propagate this certificate across the cluster so I need to include the private key to be transferred on the wire.

ayende replied:
Hm.... you are correct. I didn't take into account the fact that we need this to be cross cluster.
In this case, I think we can just put the cert (with private key) in the db record and be done with it.
We already have similarly sensitive stuff with connection strings, and they go to the same location.

ayende commented on src/Raven.Server/Web/System/TcpConnectionInfoHandler.cs:
```diff
@@ -19,5 +27,116 @@ public Task Get()
             return Task.CompletedTask;
         }
+
+        [RavenAction("/info/tcp/pull", "GET", AuthorizationStatus.UnauthenticatedClients)]
```
Let's use a better name here: `info/remote-task/topology`, for example
This won't be the only usage of the endpoint

karmeli87 replied:
Done.

ayende commented on src/Raven.Server/Web/System/TcpConnectionInfoHandler.cs:
```diff
@@ -19,5 +27,116 @@ public Task Get()
+
+            var nodes = GetResponsibleNode(databaseName, databaseGroupId, definitionName);
+
+            // TODO: cache the tcp info for each node?
```
No, there is no need for that. The number of calls you are likely to get is minuscule.

ayende commented on src/Raven.Server/Web/System/TcpConnectionInfoHandler.cs:
```diff
@@ -19,5 +27,116 @@ public Task Get()
+
+        [RavenAction("/info/tcp/pull", "GET", AuthorizationStatus.UnauthenticatedClients)]
+        public async Task GetPullReplicationInfo()
+        {
```
Security check need to be right here, front and center so it would be obvious that is is happening

ayende commented on src/Raven.Server/Web/System/TcpConnectionInfoHandler.cs:
```diff
@@ -19,5 +27,116 @@ public Task Get()
+            }
+
+            await Task.WhenAll(tasks);
+            foreach (var task in tasks)
```
Shouldn't this be in the proper order? Right now the current node tag is always first.

karmeli87 replied:
The `GetResponsibleNode` take care of the proper order by executing `WhoseTaskIsIt` in a loop and removed the selected node every iteration.

ayende replied:
Yes, _except_ that after that has been called, you call to `output.Add` directly, so it isn't in the right order

ayende commented on src/Raven.Server/Web/System/TcpConnectionInfoHandler.cs:
```diff
@@ -19,5 +27,116 @@ public Task Get()
+            }
+        }
+
+        private async Task<TcpConnectionInfo> GetNodeTcpInfo(string nodeUrl, string databaseName)
```
Why are you already getting the `TcpConnectionInfo` here? Let the client actually do that.
The response from here should be just the node urls as `http/s`

ayende replied:
The reason this concerns me is that you are using the internal channels, and the remote server might be using a different network.

karmeli87 replied:
The reason here is that `/info/tcp` require `ValidUser` which the edge don't have.

ayende replied:
So have another endpoint that would do the validation properly.
The problem of getting internal ip is something we want to avoid

ayende commented on src/Raven.Server/Web/System/TcpConnectionInfoHandler.cs:
```diff
@@ -19,5 +27,116 @@ public Task Get()
+            {
+                var pullReplication = ServerStore.Cluster.ReadPullReplicationDefinition(databaseName, definitionName, context);
+                var cert = HttpContext.Connection.ClientCertificate;
+                pullReplication.Validate(cert?.Thumbprint);
```
This is wrong, because it presume too many states.
You need to account for: 
* http node
* https with a known certificate that doesn't have access to this db
* https with a known certificate that _has_ access to this db
* https with a known admin certificate
* https with an unknown certificate that is explicitly allowed, etc
* using a database that doesn't exists.

This gives you a good way to check (with no auth) if a db exists.

ppekrol commented on src/Raven.Client/Documents/Commands/GetPullReplicationInfoCommand.cs:
```diff
@@ -0,0 +1,55 @@
+
+namespace Raven.Client.Documents.Commands
+{
+    public class GetPullReplicationInfoCommand : RavenCommand<TcpConnectionInfo[]>
```
internal

ppekrol commented on src/Raven.Client/Documents/Commands/GetPullReplicationInfoCommand.cs:
```diff
@@ -0,0 +1,55 @@
+        private readonly string _databaseGroupId;
+        private readonly string _name;
+
+        public GetPullReplicationInfoCommand(string remoteDatabase, string databaseGroupId, string name)
```
if all of them are required, check for nulls

ppekrol commented on src/Raven.Client/Documents/Operations/Replication/UpdatePullReplicationDefinitionOperation.cs:
```diff
@@ -0,0 +1,75 @@
+
+namespace Raven.Client.Documents.Operations.Replication
+{
+    public class UpdatePullReplicationDefinitionOperation : IMaintenanceOperation<ModifyOngoingTaskResult>
```
PutPullReplicationOperation?

ppekrol commented on src/Raven.Client/Documents/Replication/Messages/ReplicationLatestEtagRequest.cs:
```diff
@@ -12,4 +14,17 @@ public class ReplicationLatestEtagRequest
+    {
+        public bool PullReplication { get; set; }
+
+        public string PullReplicationDefinition { get; set; }
```
string?

ppekrol commented on src/Raven.Client/Http/RequestExecutor.cs:
```diff
@@ -224,6 +224,28 @@ public static RequestExecutor CreateForSingleNodeWithoutConfigurationUpdates(str
             return executor;
         }
 
+        public static RequestExecutor CreateForFixedTopology(string[] initialUrls, string databaseName, X509Certificate2 certificate, DocumentConventions conventions)
```
internal

ppekrol commented on src/Raven.Client/ServerWide/Tcp/TcpConnectionHeader.cs:
```diff
@@ -180,6 +206,7 @@ public class ReplicationFeatures
                 public bool MissingAttachments;
                 public bool Counters;
                 public bool ClusterTransaction;
+                public bool SupportPullReplication;
```
PullReplication

ppekrol commented on src/Raven.Client/ServerWide/Tcp/TcpConnectionHeader.cs:
```diff
@@ -37,12 +62,13 @@ public enum OperationTypes
         public static readonly int ReplicationBaseLine = 31;
         public static readonly int ReplicationAttachmentMissing = 40_300;
         public static readonly int ReplicationAttachmentMissingVersion41 = 41_300;
+        public static readonly int ReplicationWithPullOption = 42_000;
```
42_300

ppekrol commented on src/Raven.Server/Documents/Handlers/PullReplicationHandler.cs:
```diff
@@ -0,0 +1,36 @@
+            ServerStore.LicenseManager.AssertCanAddPullReplication();
+            await DatabaseConfigurations((_, databaseName, blittableJson) =>
+                {
+                    var pullReplication = JsonDeserializationClient.PullReplicationDefinition(blittableJson);
```
Maybe do some validation?

ppekrol commented on src/Raven.Server/Documents/Replication/ReplicationLoader.cs:
```diff
@@ -747,6 +853,38 @@ private TcpConnectionInfo GetConnectionInfo(ReplicationNode node, bool external)
+                    var certificate = _server.Cluster.Read(context, key);
+                    if (certificate == null)
+                    {
+                        return _server.Server.Certificate.Certificate;
```
Is this correct? When specified key does not exist then we want to use the server certificate? If yes, then probably we should check users intention with `string.IsNullOfWhiteSpace(CertificateThumbprint)` ?

ayende replied:
That is a really good point. If the user specified a value, it _must_ exists or generate an error

ppekrol commented on src/Raven.Server/ServerWide/ClusterCommandsVersionManager.cs:
```diff
@@ -95,8 +95,9 @@ public static class ClusterCommandsVersionManager
-        };
+            [nameof(DeleteSorterCommand)] = Base42CommandsVersion,
+	        [nameof(UpdatePullReplicationCommand)] = Base42CommandsVersion
+ };
```
:)

ppekrol commented on src/Raven.Server/ServerWide/ClusterStateMachine.cs:
```diff
@@ -1431,6 +1433,39 @@ public BlittableJsonReaderObject ReadRawDatabase<T>(TransactionOperationContext<
+
+        public PullReplicationDefinition ReadPullReplicationDefinition(string database, string definitionName, TransactionOperationContext context)
+        {
+            var databaseRecord = ReadRawDatabase(context, database, out _);
```
put this in 'using' ?

ppekrol commented on src/Raven.Server/ServerWide/ClusterStateMachine.cs:
```diff
@@ -1431,6 +1433,39 @@ public BlittableJsonReaderObject ReadRawDatabase<T>(TransactionOperationContext<
+            using (var rawDatabaseRecord = ReadRawDatabase(context, name, out _))
+            {
+                rawDatabaseRecord.TryGet(nameof(DatabaseRecord.Topology), out BlittableJsonReaderObject topology);
+                using (topology)
```
No need if you are disposing the parent (rawDatabaseRecord) ?

ppekrol commented on src/Raven.Server/ServerWide/ClusterStateMachine.cs:
```diff
@@ -1431,6 +1433,39 @@ public BlittableJsonReaderObject ReadRawDatabase<T>(TransactionOperationContext<
+        {
+            using (var rawDatabaseRecord = ReadRawDatabase(context, name, out _))
+            {
+                rawDatabaseRecord.TryGet(nameof(DatabaseRecord.Topology), out BlittableJsonReaderObject topology);
```
Can this return false? Exception maybe in that case?

ppekrol commented on src/Raven.Server/ServerWide/Commands/UpdatePullReplicationCommand.cs:
```diff
@@ -0,0 +1,30 @@
+{
+    public class UpdatePullReplicationCommand : UpdateDatabaseCommand
+    {
+        public PullReplicationDefinition PullReplicationDefinition;
```
rename to Definition

ppekrol commented on src/Raven.Server/Web/Authentication/AdminCertificatesHandler.cs:
```diff
@@ -990,7 +1019,7 @@ private static void ValidatePermissions(CertificateDefinition certificate, Serve
             {
                 if (string.IsNullOrWhiteSpace(kvp.Key))
-                    throw new ArgumentException("Error in permissions in the certificate definition, database name is empty");
+                    throw new ArgumentException("Error in permissions in the certificate centralDefinition, database name is empty");
```
centralDefinition ?

ppekrol commented on test/FastTests/Client/BulkInserts.cs:
```diff
@@ -17,6 +17,7 @@ public class BulkInserts : RavenTestBase
     {
         [Theory]
         [InlineData(false)]
+        [InlineData(true)]
```
SSL one was moved to SlowTests. This can be removed.

ppekrol commented on src/Raven.Client/Documents/Operations/Replication/ExternalReplication.cs:
```diff
@@ -16,6 +16,8 @@ public class ExternalReplication : ReplicationNode, IDatabaseTask, IDynamicJsonV
         public string MentorNode;
         public TimeSpan DelayReplicationFor;
 
+        public PullReplicationAsEdgeSettings PullReplicationAsEdgeSettings;
```
can we rename to Options? In most cases we are using Options.

ppekrol commented on src/Raven.Client/Documents/Operations/Replication/ExternalReplication.cs:
```diff
@@ -105,4 +108,31 @@ public string GetTaskName()
+
+    public class PullReplicationAsEdgeSettings
+    {
+        public bool PullReplicationEnabled = true;
```
Just Enabled?

ppekrol commented on src/Raven.Client/ServerWide/DatabaseRecord.cs:
```diff
@@ -61,7 +61,9 @@ public DatabaseRecord(string databaseName)
-        public List<ExternalReplication> ExternalReplications = new List<ExternalReplication>(); // Watcher only receives (slave)
+        public List<ExternalReplication> ExternalReplications = new List<ExternalReplication>();
+
+        public Dictionary<string, PullReplicationDefinition> PullReplicationDefinition = new Dictionary<string, PullReplicationDefinition>(); 
```
PullReplications

ppekrol commented on src/Raven.Server/Documents/Replication/OutgoingReplicationHandler.cs:
```diff
@@ -118,237 +134,156 @@ public string GetNode()
+            using (_parent._server.ContextPool.AllocateOperationContext(out TransactionOperationContext context))
+            using (context.OpenReadTransaction())
+            {
+                var record = _parent.LoadDatabaseRecord();
```
If this happens often, then we can use ReadRawRecord?

ayende replied:
Probably a good idea

karmeli87 committed:
RavenDB-11352 Pull Replication

addressing most of the PR comments

ayende commented on src/Raven.Client/Documents/Operations/Replication/FeatureTaskDefinition.cs:
```diff
@@ -23,6 +24,10 @@ public virtual DynamicJsonValue ToJson()
             DynamicJsonArray certs = null;
             if (Certificates != null)
             {
+                foreach (var certificate in Certificates)
```
You can assume that this is valid here.

karmeli87 replied:
Remove this, we don't need to validate this

ayende commented on src/Raven.Client/Documents/Operations/Replication/FeatureTaskDefinition.cs:
```diff
@@ -1,13 +1,14 @@
     public abstract class FeatureTaskDefinition : IDynamicJsonValueConvertible
     {
-        public List<string> Certificates; // thumbprint
+        public List<string> Certificates; // base64
```
Change to Dictionary<thubmprint, base64 cert>

karmeli87 replied:
Should be a dictionary<thumb, base 64 cert>

ayende commented on src/Raven.Client/Documents/Operations/Replication/FeatureTaskDefinition.cs:
```diff
@@ -33,21 +38,35 @@ public virtual DynamicJsonValue ToJson()
             {
-                throw new AuthorizationException($"Certificate with the thumbprint {thumbprint} was not found for pull replication '{Name}'");
+                var cert = new X509Certificate2(Convert.FromBase64String(certificate));
+                if (cert.Thumbprint == thumbprint)
```
Now that you have a match, you need to note that you _don't_ need to validate the certificate here.
Why? Because it was already validated (expiration, etc) by the calling code.

ayende commented on src/Raven.Client/Documents/Operations/Replication/PullReplicationDefinition.cs:
```diff
@@ -34,5 +34,13 @@ public override DynamicJsonValue ToJson()
             return djv;
         }
+
+        public static string GetPrefix(string databaseName)
```
remove

karmeli87 replied:
Remove

ayende commented on src/Raven.Client/ServerWide/Operations/Certificates/PutFeatureCertificateOperation.cs:
```diff
@@ -2,13 +2,14 @@
 namespace Raven.Client.ServerWide.Operations.Certificates
 {
-    public class PutFeatureCertificateOperation : IServerOperation
+    public class PutFeatureCertificateOperation : IMaintenanceOperation
```
remove

karmeli87 replied:
Remove

karmeli87 commented on src/Raven.Client/Documents/Operations/Replication/FeatureTaskDefinition.cs:
```diff
@@ -33,21 +38,35 @@ public virtual DynamicJsonValue ToJson()
+            foreach (var certificate in Certificates)
             {
-                throw new AuthorizationException($"Certificate with the thumbprint {thumbprint} was not found for pull replication '{Name}'");
+                var cert = new X509Certificate2(Convert.FromBase64String(certificate));
```
Just validating the thumbprint is fine, need to note here that we rely on calling code to validate the actual cert (expiration, etc)

karmeli87 commented on src/Raven.Server/Documents/Handlers/PullReplicationHandler.cs:
```diff
@@ -32,5 +42,62 @@ public async Task DefinePullReplication()
                 }, statusCode: HttpStatusCode.Created);
         }
+
+        [RavenAction("/databases/*/admin/certificates/feature", "PUT", AuthorizationStatus.Operator)]
```
Remove

karmeli87 commented on src/Raven.Server/Documents/Handlers/PullReplicationHandler.cs:
```diff
@@ -32,5 +42,62 @@ public async Task DefinePullReplication()
+        }
+
+        [RavenAction("/databases/*/info/remote-task/tcp", "GET", AuthorizationStatus.RestrictedAccess)]
+        public Task GetRemoteTaskTcp()
```
You cannot do this. This leaks the existence of a database by the fact that you are using `DatabaseRequestHandler`
You need to use a `RequestHandler` and accept the database name externally.

karmeli87 commented on src/Raven.Server/Documents/Handlers/PullReplicationHandler.cs:
```diff
@@ -32,5 +42,62 @@ public async Task DefinePullReplication()
+        public static void Authenticate(HttpContext httpContext, ServerStore serverStore, string database, string remoteTask)
+        {
+            // TODO: think of some use cases like:
+            // 1. HTTP, but we defined cert.
```
Should allow, this is already handled externally and not here.
If we don't have HTTPS, we always allow.

karmeli87 commented on src/Raven.Server/Documents/Handlers/PullReplicationHandler.cs:
```diff
@@ -32,5 +42,62 @@ public async Task DefinePullReplication()
+        {
+            // TODO: think of some use cases like:
+            // 1. HTTP, but we defined cert.
+            // 2. HTTPs, but no cert is defined.
```
This is invalid, and even if we have restricted here, we should abort it early

karmeli87 commented on src/Raven.Server/Documents/Handlers/PullReplicationHandler.cs:
```diff
@@ -32,5 +42,62 @@ public async Task DefinePullReplication()
+            // TODO: think of some use cases like:
+            // 1. HTTP, but we defined cert.
+            // 2. HTTPs, but no cert is defined.
+            // 3. HTTPs with some cert, but got the server cert.
```
If the cert has access, we get access.

karmeli87 commented on src/Raven.Server/Documents/Handlers/PullReplicationHandler.cs:
```diff
@@ -32,5 +42,62 @@ public async Task DefinePullReplication()
+                    // we can trust this certificate
+                    return;
+
+                case null:
```
This is invalid

karmeli87 commented on src/Raven.Server/Routing/RequestRouter.cs:
```diff
@@ -239,6 +239,12 @@ private bool TryAuthorize(RouteInformation route, HttpContext context, DocumentD
                         case RavenServer.AuthenticationStatus.NotYetValid:
                         case RavenServer.AuthenticationStatus.None:
                         case RavenServer.AuthenticationStatus.UnfamiliarCertificate:
```
create another case and if this fails, `goto null` will handle this

karmeli87 commented on src/Raven.Server/Web/Authentication/AdminCertificatesHandler.cs:
```diff
@@ -198,175 +196,169 @@ public static void WriteCertificateAsPem(string name, byte[] rawBytes, string ex
             return PutCertificate(Constants.Certificates.Prefix, includePrivateKey: false);
         }
 
-        private async Task PutCertificate(string prefix, bool includePrivateKey)
```
nope

karmeli87 commented on src/Raven.Server/Web/RequestHandler.cs:
```diff
@@ -537,5 +541,198 @@ protected void SetupCORSHeaders()
             SetupCORSHeaders(HttpContext);
         }
+
+        protected async Task PutCertificate(string prefix, bool includePrivateKey)
```
nope

karmeli87 commented on src/Raven.Server/Web/RequestHandler.cs:
```diff
@@ -537,5 +541,198 @@ protected void SetupCORSHeaders()
+            }
+        }
+
+        public static async Task PutCertificateCollectionInCluster(
```
nope

karmeli87 committed:
RavenDB-11352 adderss the PR comments

still to-do:

- Collect the info who is connected to the centeral.
- Write tests to verify the correct behavior when we disable/enbale.

ayende commented on src/Raven.Client/Documents/Operations/Replication/FeatureTaskDefinition.cs:
```diff
@@ -0,0 +1,69 @@
+            if (string.IsNullOrEmpty(thumbprint))
+            {
+                if (Certificates == null || Certificates.Count == 0)
+                    return true;
```
This is wrong.  There no scenario where we have an "anonymous" pull replication, which is this is what this looks like.
If `Certificates` is null / empty, we return a negative result and do not allow access.

Note that this is different from the behavior when we use a known client certificate that already have access to this node / database.

ayende commented on src/Raven.Server/Documents/Replication/OutgoingReplicationHandler.cs:
```diff
@@ -118,237 +132,130 @@ public string GetNode()
-                        throw new InvalidOperationException($"The database record for {_parent.Database.Name} does not exist?!");
+        private void Replication()
+        {
+            NativeMemory.EnsureRegistered();
```
remove this line

ayende commented on src/Raven.Server/Documents/Replication/OutgoingReplicationHandler.cs:
```diff
@@ -425,6 +332,225 @@ void HandleException(Exception e)
+                                {
+                                    _parent.EnsureNotDeleted(dest.NodeTag);
+                                }
+
```
Probably need to validate that this is still valid here. 
Need to revalidate if the pull replication def changed (especially if just a cert was removed)

ayende commented on src/Raven.Server/Documents/Replication/ReplicationLoader.cs:
```diff
@@ -160,7 +164,107 @@ public ReplicationLoader(DocumentDatabase database, ServerStore server)
+            void RetryPullReplication(IncomingReplicationHandler incomingReplicationHandler, Exception exception)
+            {
+                // if the stream closed, it is our duty to reconnect
+                incomingReplicationHandler.Failed -= RetryPullReplication;
```
what about retries, backoff strategy, etc?

ayende commented on src/Raven.Server/RavenServer.cs:
```diff
@@ -1563,6 +1583,22 @@ private bool TryAuthorize(RavenConfiguration configuration, Stream stream, TcpCo
+                        using (ctx.OpenReadTransaction())
+                        {
+                            var pullReplicationDefinition = ServerStore.Cluster.ReadPullReplicationDefinition(header.DatabaseName, info.AuthorizationFor, ctx);
+                            if (pullReplicationDefinition.CanAccess(certificate.Thumbprint, out var err))
```
What if the pull replication doesn't exists?
Note that in this case we must provide the same error as if the cert doesn't exists.
What error do we get from missing db in this case, for that matter?

ayende commented on src/Raven.Server/ServerWide/ServerStore.cs:
```diff
@@ -1329,58 +1328,20 @@ public async Task<(long, object)> ModifyPeriodicBackup(TransactionOperationConte
                     break;
                 case EtlType.Sql:
-                    var sqlEtl = JsonDeserializationCluster.SqlEtlConfiguration(etlConfiguration);
-                    sqlEtl.Validate(out var sqlEtlErr, validateName: false, validateConnection: false);
```
Where is the validation now?

ayende commented on src/Raven.Server/ServerWide/ServerStore.cs:
```diff
@@ -1329,58 +1328,20 @@ public async Task<(long, object)> ModifyPeriodicBackup(TransactionOperationConte
             {
                 case EtlType.Raven:
-                    var rvnEtl = JsonDeserializationCluster.RavenEtlConfiguration(etlConfiguration);
-                    rvnEtl.Validate(out var rvnEtlErr, validateName: false, validateConnection: false);
```
Where is the validation now?

ayende commented on src/Raven.Server/Web/Authentication/AdminCertificatesHandler.cs:
```diff
@@ -999,5 +997,14 @@ private static void ValidatePermissions(CertificateDefinition certificate, Serve
                     throw new ArgumentException($"Error in permissions in the certificate definition, invalid access {kvp.Value} for database {kvp.Key}");
             }
         }
+
```
Do you actually need to do any of the changes to this file now?

karmeli87 replied:
Don't think so. But since it was changed in different commits I'm not sure I can simply revert the file in this PR.

ayende replied:
yes

ayende commented on src/Raven.Server/Web/System/TcpConnectionInfoHandler.cs:
```diff
@@ -19,5 +30,139 @@ public Task Get()
+                    {
+                        var pullReplication = serverStore.Cluster.ReadPullReplicationDefinition(database, remoteTask, context);
+                        var cert = httpContext.Connection.ClientCertificate;
+                        if (pullReplication.CanAccess(cert?.Thumbprint, out var err) == false)
```
what happens if the db doesn't exists? What happens if the pull replication doesn't exists?

karmeli87 replied:
If the db doesn't exists `throw new DatabaseDoesNotExistException($"The database '{database}' doesn't exists.")`
If the pull replication doesn't exists `throw new InvalidOperationException($"Pull replication with the name '{definitionName}' isn't defined for the database '{database}'.");`

ayende replied:
Both of which leak information that you may not be allowed to see. 
We need to make sure that regardless of the method you use, if you don't have access to the db / pull replication, we reply with the same message:

"Cannot connect to '{pullReplicationName}' on '{database}'. The database or task may not exists or you don't have the credentials to access it"`

ayende commented on src/Raven.Server/Web/System/TcpConnectionInfoHandler.cs:
```diff
@@ -19,5 +30,139 @@ public Task Get()
+
+            switch (feature.Status)
+            {
+                case RavenServer.AuthenticationStatus.Allowed:
```
This isn't true. The certificate is allowed to _some_ database, not necessarily this one.

karmeli87 committed:
RavenDB-11352 unify authorization exception message.

karmeli87 committed:
RavenDB-11352 Make pull replication as a separate property on the edge database record

- allow edit pull replicaiton on edge/central
- collect info about the pull replication from edge/central

karmeli87 committed:
RavenDB-11352 add enable/disable for pull replication

ayende commented on src/Raven.Server/Web/System/TcpConnectionInfoHandler.cs:
```diff
@@ -45,6 +46,9 @@ public Task GetRemoteTaskTopology()
             using (context.OpenReadTransaction())
             {
                 var pullReplication = ServerStore.Cluster.ReadPullReplicationDefinition(database, remoteTask, context);
+                if (pullReplication.Disabled)
```
You are only checking this _when the connection is started_, you need to also check it whenever the configuration changes and abort any existing connections

karmeli87 replied:
This is done under `HandleCentralPullReplication`.

karmeli87 commented:
Closing this PR. 

Will create a new one with the new terms `Hub` and `Sink`
